kopf/clients/auth.py

```diff
@@ -1,42 +1,125 @@
 import logging
+from typing import Optional
 
-import kubernetes
+import pykube
+import requests
 import urllib3.exceptions
 
 logger = logging.getLogger(__name__)
 
+# Set in login(), consumed in get_pykube_cfg() and all API calls.
+_pykube_cfg: Optional[pykube.KubeConfig] = None
+
 
 class LoginError(Exception):
     """ Raised when the operator cannot login to the API. """
 
 
-def login():
+class AccessError(Exception):
+    """ Raised when the operator cannot access the cluster API. """
+
+
+def login(verify=False):
     """
-    Login the the Kubernetes cluster, locally or remotely.
+    Login to Kubernetes cluster, locally or remotely.
+
+    Keep the logged in state or config object in the global variables,
+    so that it can be available for future calls via the same function call.
+
+    Automatic refresh/reload of the tokens or objects also should be done here.
     """
 
-    # Configure the default client credentials for all possible environments.
+    # Pykube login is mandatory. If it fails, the framework will not run at all.
+    try:
+        import pykube
+    except ImportError:
+        raise  # mandatory
+    else:
+        login_pykube(verify=verify)
+
+    # We keep the official client library auto-login only because it was
+    # an implied behavior before switching to pykube -- to keep it so (implied).
```

Comment:
Is there some kind of deprecation/removal planned for this?

Reply:
Not decided on this yet. Definitely before 1.0 release. 

The deprecation of auto-login can be done only after there is a way (read: a handler) to login the operator itself when it starts — e.g. as in #59. Now, it can be done with the module-level code, but it looks hacky.